# How Floats work (in a bit more depth)

At the fundamental level computers work by flicking on and off trillions of little switches (transistors). 

When Python receives a number (in any base) it converts that number into base 2 (binary). This is because at the fundamental level computers run on binary (1’s and 0’s get further converted into electrical current; say 2V for 1 and 0V for 0). 

So numbers start at base 10 → are converted to base 2 → and then are converted back again. Sometimes information can be lost in this process. As a result sometimes Python will give you weird answers to basic arithmetic questions. It might seem like a bug when it happens. But actually, these ‘quirks’ are just a consequence of how floats are implemented at the low level.

Let's take a look at a few examples...

In [1]:
print (0.1 + 0.1)

0.2


0.1 + 0.1 = 0.2 Okay, no surprises here. But what about the following sum:

In [2]:
print(0.1 + 0.1 + 0.1)

0.30000000000000004


This sum should equal 0.3, you don’t need to be a genius to figure that out. But what Python returns is close to right answer but nonetheless wrong. 

What happened? well 0.1 * 3 is easy for us to do in base 10 but in base 2 it’s a bit more tricky. Instead of trying to find the actual value Python "gives up" after a while and just does a bit of rounding.

Here’s another example:

In [11]:
print(0.7 + 0.7 + 0.7)
print(0.7 + 0.7 + 0.7 + 0.7)

2.0999999999999996
2.8


Here 0.7 times 3 is clearly 2.1 but Python returns 2.099. If we add an extra 0.7 to the mix Python gets the anwser correct, 0.7 times 4 does in fact equal 2.8. 

Alright, that just about wraps up this lecture on numbers. If you would like an even more in-depth explanation how floats works please [read this](https://docs.python.org/3/tutorial/floatingpoint.html). 